go/pkg/cas/upload.go

```diff
@@ -441,12 +476,67 @@ func (u *uploader) check(ctx context.Context, items []*uploadItem) error {
 }
 
 func (u *uploader) scheduleUpload(ctx context.Context, item *uploadItem) error {
-	if u.testScheduleUpload != nil {
-		return u.testScheduleUpload(ctx, item)
+	// Check if this blob can be uploaded in a batch.
+	if marshalledRequestSize(item.Digest) > int64(u.batchBundler.BundleByteLimit) {
+		// There is no way this blob can fit in a batch request.
+		// TODO(nodir): implement streaming.
+		panic("not implemented")
 	}
 
-	// TODO(nodir): implement.
-	panic("not implemented")
+	// Since this blob is small enough, just read it entirely.
+	contents, err := item.ReadAll()
+	if err != nil {
+		return errors.Wrapf(err, "failed to read the item")
+	}
+	req := &repb.BatchUpdateBlobsRequest_Request{Digest: item.Digest, Data: contents}
+	return u.batchBundler.AddWait(ctx, req, proto.Size(req))
+}
+
+// uploadBatch uploads blobs in using BatchUpdateBlobs RPC.
+func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlobsRequest_Request) error {
+	if err := u.semBatches.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semBatches.Release(1)
+
+	reqMap := make(map[digest.Digest]*repb.BatchUpdateBlobsRequest_Request, len(reqs))
+	for _, r := range reqs {
+		reqMap[digest.NewFromProtoUnvalidated(r.Digest)] = r
+	}
+
+	req := &repb.BatchUpdateBlobsRequest{
+		InstanceName: u.InstanceName,
+		Requests:     reqs,
+	}
+	return u.retry(ctx, func() error {
+		// TODO(nodir): add per-RPC timeouts.
+		res, err := u.cas.BatchUpdateBlobs(ctx, req)
```

Comment:
Why not use https://github.com/bazelbuild/remote-apis-sdks/blob/fd877b05ba2ed611f6d34e711da05917b729eff9/go/pkg/client/client.go#L821? Or is this plan once we move things out of the client package?

Reply:
Because `client` package will eventually depend on `cas` package, so we cannot have `cas` package heavily depending on `client` package. Besides, `client` package is scoped to execution - will we still want `BatchUploadBlobs` in the client package?

Reply:
I see. So I guess this question is premature, until we start using `cas` in `client` and then we will figure out how the retry policy is passed through and how RPC timeouts are set.

Reply:
the retry policy can be passed by setting `ClientConfig.RetryPolicy` today.
I expect RPC Timeouts to be similar, but it is not done yet.

Reply:
Ack.